Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1134 @@
+
+instance LieRing.intLieAlgebra : LieAlgebra ℤ L where lie_smul n x y := lie_zsmul x y n
+#align lie_ring.int_lie_algebra LieRing.intLieAlgebra
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes. (lean4#2074)
+attribute [-instance] Ring.toNonAssocRing
+
+instance : LieRingModule L (M →ₗ[R] N) where
+  bracket x f :=
+    { toFun := fun m => ⁅x, f m⁆ - f ⁅x, m⁆
+      map_add' := fun m n => by
+        simp only [lie_add, LinearMap.map_add]
+        abel
+      map_smul' := fun t m => by
+        simp only [smul_sub, LinearMap.map_smul, lie_smul, RingHom.id_apply] }
+  add_lie x y f := by
+    ext n
+    simp only [add_lie, LinearMap.coe_mk, AddHom.coe_mk, LinearMap.add_apply, LinearMap.map_add]
+    abel
+  lie_add x f g := by
+    ext n
+    simp only [LinearMap.coe_mk, AddHom.coe_mk, lie_add, LinearMap.add_apply]
+    abel
+  leibniz_lie x y f := by
+    ext n
+    simp only [lie_lie, LinearMap.coe_mk, AddHom.coe_mk, LinearMap.map_sub, LinearMap.add_apply,
+      lie_sub]
+    abel
+
+@[simp]
+theorem LieHom.lie_apply (f : M →ₗ[R] N) (x : L) (m : M) : ⁅x, f⁆ m = ⁅x, f m⁆ - f ⁅x, m⁆ :=
+  rfl
+#align lie_hom.lie_apply LieHom.lie_apply
+
+instance : LieModule R L (M →ₗ[R] N)
+    where
+  smul_lie t x f := by
+    ext n
+    simp only [smul_sub, smul_lie, LinearMap.smul_apply, LieHom.lie_apply, LinearMap.map_smul]
+  lie_smul t x f := by
+    ext n
+    simp only [smul_sub, LinearMap.smul_apply, LieHom.lie_apply, lie_smul]
+
+end BasicProperties
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes. (lean4#2074)
+attribute [-instance] Ring.toNonAssocRing
+
+/-- A morphism of Lie algebras is a linear map respecting the bracket operations. -/
+structure LieHom (R L L': Type _) [CommRing R] [LieRing L] [LieAlgebra R L]
+  [LieRing L'] [LieAlgebra R L'] extends L →ₗ[R] L' where
+  /-- A morphism of Lie algebras is compatible with brackets. -/
+  map_lie' : ∀ {x y : L}, toFun ⁅x, y⁆ = ⁅toFun x, toFun y⁆
+#align lie_hom LieHom
+
+@[inherit_doc]
+notation:25 L " →ₗ⁅" R:25 "⁆ " L':0 => LieHom R L L'
+
+namespace LieHom
+
+variable {R : Type u} {L₁ : Type v} {L₂ : Type w} {L₃ : Type w₁}
+
+variable [CommRing R]
+
+variable [LieRing L₁] [LieAlgebra R L₁]
+
+variable [LieRing L₂] [LieAlgebra R L₂]
+
+variable [LieRing L₃] [LieAlgebra R L₃]
+
+attribute [coe] LieHom.toLinearMap
+
+instance : Coe (L₁ →ₗ⁅R⁆ L₂) (L₁ →ₗ[R] L₂) :=
+  ⟨LieHom.toLinearMap⟩
+
+/-- Coercion to a function. -/
+@[coe] def toFun' (f : L₁ →ₗ⁅R⁆ L₂) := f.toFun
+
+/-- see Note [function coercion] -/
+instance : CoeFun (L₁ →ₗ⁅R⁆ L₂) fun _ => L₁ → L₂ :=
+  ⟨toFun'⟩
+
+/-- See Note [custom simps projection]. We need to specify this projection explicitly in this
+  case, because it is a composition of multiple projections. -/
+def Simps.apply (h : L₁ →ₗ⁅R⁆ L₂) : L₁ → L₂ :=
+  h
+#align lie_hom.simps.apply LieHom.Simps.apply
+
+-- Porting note: Does not work
+-- initialize_simps_projections LieHom (toLinearMap_toFun → apply)
```

Comment:
I could not get this to work. Furthermore for reasons I don't understand `LieHom.toLinearMap_toAddHom_toFun` does not exist.